File: src/StackAndQueue/ValidParentheses.java
package StackAndQueue;

import java.util.Stack;

// https://leetcode.com/problems/valid-parentheses/
public class ValidParentheses {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(')
                stack.push(')');
            else if (c == '{')
                stack.push('}');
            else if (c == '[')
                stack.push(']');
            else if (stack.isEmpty() || stack.pop() != c)
                return false;
        }
        return stack.isEmpty();
    }

    // Driver method
    public static void main(String[] args) throws java.lang.Exception {
        String input = "[]{}";
        ValidParentheses v = new ValidParentheses();
        System.out.println("Is valid: " + v.isValid(input));
    }
}

File: py/essential/ValidParentheses.py
# https://leetcode.com/problems/valid-parentheses/
# Stack
# Blind
# Grind
# Neet


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dict = {")": "(", "}": "{", "]": "["}

        for letter in s:
            if letter in dict.values():  # {[(
                stack.append(letter)
            elif letter in dict.keys():
                if not stack or stack.pop() != dict[letter]:
                    return False
            else:
                return False
        return not stack


s = Solution()
input = "{(){}[]}"
print(s.isValid(input))
